File: tests/test_vulnerable.js
const express = require('express');
const mysql = require('mysql');
const fs = require('fs');
const path = require('path');

const app = express();

function unsafeDbQuery(userId, searchTerm) {
    const connection = mysql.createConnection({
        host: 'localhost',
        user: 'dbuser',
        password: 'dbpass',
        database: 'myapp'
    });
    
    // SQL injection vulnerability - string concatenation
    const query = `SELECT * FROM users WHERE id = ${userId} AND name LIKE '%${searchTerm}%'`;
    
    connection.query(query, (error, results) => {
        if (error) throw error;
        console.log(results);
    });
    
    connection.end();
}

function vulnerableFileAccess(userFilename) {
    // Path traversal vulnerability
    const filePath = path.join('/uploads', userFilename);
    
    // No validation allows ../../../etc/passwd
    try {
        return fs.readFileSync(filePath, 'utf8');
    } catch (err) {
        return null;
    }
}

app.get('/profile', (req, res) => {
    const username = req.query.name;
    
    // XSS vulnerability - direct HTML injection
    const html = `<div>Welcome ${username}!</div>`;
    res.send(html);
});

function weakAuthentication(password) {
    // Authentication bypass vulnerability
    if (password === '') {
        return true; // Empty password bypass
    }
    
    // Hardcoded password
    if (password === 'admin123') {
        return true;
    }
    
    return false;
}

function dangerousEval(userCode) {
    // Code injection vulnerability
    try {
        return eval(userCode);
    } catch (e) {
        return 'Error executing code';
    }
}

function templateInjection(userInput) {
    // Template injection vulnerability
    const template = `Hello ${userInput}!`;
    
    // Using eval with template strings
    return eval('`' + template + '`');
}

// React-like XSS vulnerability
function renderUserContent(userHtml) {
    return {
        __html: userHtml // dangerouslySetInnerHTML equivalent
    };
}

app.listen(3000, () => {
    console.log('Server running on port 3000');
});

File: tests/test_vulnerable.py
import sqlite3
from flask import Flask, request, render_template_string
import os

app = Flask(__name__)

def unsafe_login(username, password):
    """Vulnerable login function with SQL injection."""
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    
    # SQL Injection vulnerability - string concatenation
    query = "SELECT * FROM users WHERE username='" + username + "' AND password='" + password + "'"
    cursor.execute(query)
    
    result = cursor.fetchone()
    conn.close()
    
    if result:
        return True
    return False

def vulnerable_file_read(filename):
    """Path traversal vulnerability."""
    # No path validation - allows ../../../etc/passwd
    file_path = "/uploads/" + filename
    
    try:
        with open(file_path, 'r') as f:
            return f.read()
    except:
        return None

@app.route('/search')
def vulnerable_search():
    """XSS vulnerability in template."""
    search_term = request.args.get('q', '')
    
    # XSS vulnerability - unescaped user input in template
    template = f"<h1>Search results for: {search_term}</h1>"
    return render_template_string(template)

def weak_auth_check(token):
    """Authentication bypass vulnerability."""
    # Hardcoded bypass - always returns True for admin
    if token == "admin123":
        return True
    
    # Weak token validation
    if not token or token == "":
        return False
    
    return True

def dangerous_eval(user_input):
    """Code injection vulnerability."""
    # Direct eval of user input
    try:
        result = eval(user_input)
        return result
    except:
        return "Error"

def format_string_sql(user_id, action):
    """SQL injection via f-string."""
    conn = sqlite3.connect('logs.db')
    cursor = conn.cursor()
    
    # SQL injection via f-string
    query = f"INSERT INTO logs (user_id, action) VALUES ({user_id}, '{action}')"
    cursor.execute(query)
    conn.commit()
    conn.close()

if __name__ == "__main__":
    app.run(debug=True)